cmd/src/actions_exec_logger.go

```diff
@@ -227,9 +225,106 @@ func (a *actionLogger) DockerStepStarted(repoName string, step int, image string
 }
 
 func (a *actionLogger) DockerStepErrored(repoName string, step int, err error, elapsed time.Duration) {
+	a.progress.IncStepsComplete(1)
 	a.write(repoName, boldRed, "%s %s. (%s)\n", boldBlack.Sprintf("[Step %d]", step), err, elapsed)
 }
 
 func (a *actionLogger) DockerStepDone(repoName string, step int, elapsed time.Duration) {
+	a.progress.IncStepsComplete(1)
 	a.write(repoName, yellow, "%s Done. (%s)\n", boldBlack.Sprintf("[Step %d]", step), elapsed)
 }
+
+func (a *actionLogger) write(repoName string, c *color.Color, format string, args ...interface{}) {
+	if w, ok := a.RepoWriter(repoName); ok {
+		fmt.Fprintf(w, format, args...)
+	}
+
+	if a.verbose {
+		if len(repoName) > 0 {
+			format = fmt.Sprintf("%s -> %s", c.Sprint(repoName), format)
+		}
+		fmt.Fprintf(a.progress, format, args...)
+	}
+}
+
+type progress struct {
```

Comment:
Hmmm, I think the API of this could be better.

`progress` is a really generic term and doesn't tell me what it does. But it seems to do lots of things: act as an IO pipe, track progress, format progress, clean previously written output.

How about splitting it up? For example, `progress` could be turned into a `progressBar` which only tracks the progress and has a `String() string` method that returns the correctly formatted bar for current progress.

The logger can then take care of when to print that progress bar and when not.

Reply:
I split it up into `progress` that tracks the progress state and `progressWriter` that uses that state to print a progress bar after performing it's write.

We need to pass in two writers for stderr and stdout here:
https://github.com/sourcegraph/src-cli/blob/1cd4ce40d15bfce1369fd508640e742d606acf2a/cmd/src/actions_exec_logger.go#L163-L167

And those need access to the same overall progress state.